src/meta-srv/src/election/etcd.rs

```diff
@@ -127,6 +131,65 @@ impl Election for EtcdElection {
             .is_ok()
     }
 
+    async fn register_candidate(&self) -> Result<()> {
+        const CANDIDATE_LEASE_SECS: u64 = 600;
+        const KEEP_ALIVE_INTERVAL_SECS: u64 = CANDIDATE_LEASE_SECS / 2;
+
+        let mut lease_client = self.client.lease_client();
+        let res = lease_client
+            .grant(CANDIDATE_LEASE_SECS as i64, None)
+            .await
+            .context(error::EtcdFailedSnafu)?;
+        let lease_id = res.id();
+
+        // The register info: key is the candidate key, value is its leader value.
+        let key = self.candidate_key().into_bytes();
+        let value = self.leader_value.clone().into_bytes();
+        // Puts with the lease id
+        self.client
+            .kv_client()
+            .put(key, value, Some(PutOptions::new().with_lease(lease_id)))
+            .await
+            .context(error::EtcdFailedSnafu)?;
+
+        let (mut keeper, mut receiver) = lease_client
+            .keep_alive(lease_id)
+            .await
+            .context(error::EtcdFailedSnafu)?;
+
+        let mut keep_alive_interval =
+            tokio::time::interval(Duration::from_secs(KEEP_ALIVE_INTERVAL_SECS));
+
+        loop {
+            let _ = keep_alive_interval.tick().await;
+            keeper.keep_alive().await.context(error::EtcdFailedSnafu)?;
+
+            if let Some(res) = receiver.message().await.context(error::EtcdFailedSnafu)? {
+                if res.ttl() <= 0 {
+                    // Failed to keep alive, just break the loop.
+                    break;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn all_candidates(&self) -> Result<Vec<LeaderValue>> {
+        let key = self.candidate_root().into_bytes();
+        let res = self
+            .client
+            .kv_client()
+            .get(key, Some(GetOptions::new().with_prefix()))
+            .await
+            .context(error::EtcdFailedSnafu)?;
+
+        res.kvs()
+            .iter()
+            .map(|kv| Ok(LeaderValue(String::from_utf8_lossy(kv.value()).to_string())))
```

Comment:
A minor suggestion: we can implement `From<Vec<u8>>` for `LeaderValue` to simplify those encodings.

Reply:
Updated.